Minor doc restructuring for clarity (#553)
compintell/Mooncake.jl: docs/src/developer_documentation/reverse_mode_design.md

# Reverse-Mode Design

## Compilation process

Last checked: 09/02/2025, Julia v1.10.8 / v1.11.3, Mooncake 0.4.83.

This brief informal note was largely written by Guillaume Dalle while learning how Mooncake's internals operate for reverse-mode, in order to be able to add forwards-mode AD.
It should help readers orient themsleves when first trying to understand Mooncake's internals.

Rule building is done statically, based on types. Some methods accept values, e.g.
```julia
build_rrule(args...; debug_mode=false)
```
but these simply extract the type of all arguments and call another method of `build_rrule`.

The action happens in [`s2s_reverse_mode_ad.jl`](https://github.com/compintell/Mooncake.jl/blob/main/src/interpreter/s2s_reverse_mode_ad.jl), in particular the following method:
```julia
build_rrule(interp::MooncakeInterpreter{C}, sig_or_mi; debug_mode=false)
```
`sig_or_mi` is either a signature, such as `Tuple{typeof(foo), Float64}`, or a `Core.MethodInstance`.
Signatures are extracted from `Core.MethodInstance`s as necessary.

If a signature has a custom rule (`Mooncake.is_primitive` returns `true`), we take it, otherwise we generate the IR and differentiate it.

> The forward- and reverse-pass IRs are created by `generate_ir`.

Reviewer: ```suggestion
The forward and reverse pass IRs are created by the [`generate_ir`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/internal_docstrings/#Mooncake.generate_ir-Tuple{Mooncake.MooncakeInterpreter,%20Any}) method.
```

The `OpaqueClosure` allows going back from the IR to a callable object. More precisely we use `MistyClosure` to store the associated IR.

The `Pullback` and `DerivedRule` structs are convenience wrappers for `MistyClosure`s with some bookkeeping.

Diving one level deeper, in the following method:

```julia
generate_ir(
    interp::MooncakeInterpreter, sig_or_mi; debug_mode=false, do_inline=true
)
```

> The function `lookup_ir` calls `Core.Compiler.typeinf_ircode` on a method instance, which is a lower-level version of `Base.code_ircode`.

Reviewer: ```suggestion
The function [`lookup_ir`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/internal_docstrings/#Mooncake.lookup_ir-Tuple{Core.Compiler.AbstractInterpreter,%20Type{%3C:Tuple}}) calls `Core.Compiler.typeinf_ircode` on a method instance, which is a lower-level version of `Base.code_ircode`.
```

> The IR considered is of type `IRCode`, which is different from the `CodeInfo` returned by `@code_typed`.

Reviewer: ```suggestion
The IR considered is of type [`IRCode`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/ir_representation/#Core.Compiler.IRCode-Tuple{Mooncake.BasicBlockCode.BBCode}), which is different from the `CodeInfo` returned by `@code_typed`.
```

This format is obtained from `CodeInfo`, used to perform most optimizations in the Julia IR in the [evaluation pipeline](https://docs.julialang.org/en/v1/devdocs/eval/), then converted back to `CodeInfo`.

> The function `normalise!` is a custom pass to modify `IRCode` and make some expressions nicer to work with.

Reviewer: ```suggestion
The function [`normalise!`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/internal_docstrings/#Mooncake.normalise!-Tuple{Core.Compiler.IRCode,%20Vector{Symbol}}) is a custom pass to modify `IRCode` and make some expressions nicer to work with.
```

The possible expressions one can encountered in lowered ASTs are documented [here](https://docs.julialang.org/en/v1/devdocs/ast/#Lowered-form).

Reverse-mode specific stuff: return type retrieval, `ADInfo`, `bbcode.jl`, `zero_like_rdata.jl`. The `BBCode` structure was a convenience for IR transformation.

Beyond the [`interpreter`](https://github.com/compintell/Mooncake.jl/blob/main/src/interpreter/) folder, check out [`tangents.jl`](https://github.com/compintell/Mooncake.jl/blob/main/src/tangents.jl) for forward mode.

> `FData` and `RData` are not useful in forward mode, `Tangent` is the right representation.

Reviewer: ```suggestion
[`Tangent`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/internal_docstrings/#Mooncake.Tangent) is the correct representation required for Forward mode AD. `FData` and `RData` are not representations needed directly.
```

> For testing, `generate_test_functions` from [`test_resources.jl`](https://github.com/compintell/Mooncake.jl/blob/src/test_utils.jl) should all pass.

Reviewer: ```suggestion
For testing, all the tests got via the `generate_test_functions` method (defined in [`test_resources.jl`](https://github.com/compintell/Mooncake.jl/blob/1894b2f23916091d5022134db0af61a75c1035ee/src/test_resources.jl#L655) must pass.
```

- `ir.stmts` is a `Core.Compiler.InstructionStream`. This represents a sequence of instructions via 5 vectors of the same length:
  - `stmts.stmt` is a vector of expressions (or other IR node types), see [AST docs](https://docs.julialang.org/en/v1/devdocs/ast/#Lowered-form)
  - `stmts.type` is a vector of types for the left-hand side of the assignment
  - three others
- `ir.cfg` is the Control Flow Graph of type `Core.Compiler.CFG`
- `ir.meta` is metadata, not important
- `ir.new_nodes` is an optimization buffer, not important
- `ir.sptypes` is for type parameters of the called function

We must maintain coherence between the various components of `IRCode` (especially `ir.stmts` and `ir.cfg`). That is the reason behind `BBCode`, to make coherence easier.
We can deduce the CFG from the statements but not the other way around: it's only composed of blocks of statement indices.
In forward mode we shouldn't have to modify anything but `ir.stmts`.
Do line by line transformation of the statements and then possibly refresh the CFG.

> Example of line-by-line transformations are in `make_ad_stmts!`.

Reviewer: ```suggestion
Examples of line-by-line transformations are as defined in [`Mooncake.make_ad_stmts!`](https://compintell.github.io/Mooncake.jl/dev/developer_documentation/internal_docstrings/#Mooncake.make_ad_stmts!).
```
